Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::StringView

cdumez opened the pull request:
#### 96f2e5485527cd11ca9b9725bdd5b2cc814b5b20
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::StringView
<a href="https://bugs.webkit.org/show_bug.cgi?id=283032">https://bugs.webkit.org/show_bug.cgi?id=283032</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/text/StringView.cpp:
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::span8 const):
(WTF::StringView::span16 const):
(WTF::StringView::CodePoints::Iterator::Iterator):
(WTF::StringView::CodeUnits::Iterator::operator* const):
(WTF::findCommon):
(WTF::endsWith):
(WTF::StringView::unsafeCharacterAt const): Deleted.

Canonical link: <a href="https://commits.webkit.org/286538@main">https://commits.webkit.org/286538@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bee1d3fce39525a4a7eff261fb0d9d73805947c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76256 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/29158 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80764 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27528 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64430 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3582 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59802 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17930 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49694 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65491 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/40144 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/47092 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22977 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25850 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/69435 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68220 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23312 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/82221 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75532 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3628 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2367 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68019 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3782 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65463 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67331 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11290 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9396 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97786 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11801 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3576 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21397 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3599 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/7028 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/5357 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -1302,12 +1292,12 @@ inline bool endsWith(StringView reference, StringView suffix)
         if (suffix.is8Bit())
-            return equal(reference.span8().data() + startOffset, suffix.span8());
-        return equal(reference.span8().data() + startOffset, suffix.span16());
+            return equal(reference.span8().subspan(startOffset).data(), suffix.span8());
+        return equal(reference.span8().subspan(startOffset).data(), suffix.span16());
     }
     if (suffix.is8Bit())
-        return equal(reference.span16().data() + startOffset, suffix.span8());
-    return equal(reference.span16().data() + startOffset, suffix.span16());
+        return equal(reference.span16().subspan(startOffset).data(), suffix.span8());
+    return equal(reference.span16().subspan(startOffset).data(), suffix.span16());
 }
```
Shouldn’t we have equal functions that take two spans?

cdumez replied:
Yes, I never finished that refactoring. I need to revive my old PR at https://github.com/WebKit/WebKit/pull/28778.

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER_USAGE in WTF::StringView
https://bugs.webkit.org/show_bug.cgi?id=283032

Reviewed by Darin Adler.

* Source/WTF/wtf/text/StringView.cpp:
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::span8 const):
(WTF::StringView::span16 const):
(WTF::StringView::CodePoints::Iterator::Iterator):
(WTF::StringView::CodeUnits::Iterator::operator* const):
(WTF::findCommon):
(WTF::endsWith):
(WTF::StringView::unsafeCharacterAt const): Deleted.

Canonical link: https://commits.webkit.org/286538@main
